src/java/main/org/apache/zookeeper/server/WatchManager.java

```diff
@@ -97,23 +102,44 @@ synchronized void removeWatcher(Watcher watcher) {
     Set<Watcher> triggerWatch(String path, EventType type, Set<Watcher> supress) {
         WatchedEvent e = new WatchedEvent(type,
                 KeeperState.SyncConnected, path);
-        HashSet<Watcher> watchers;
+        Set<Watcher> watchers = new HashSet<>();
+        PathIterator pathIterator = new PathIterator(path);
         synchronized (this) {
-            watchers = watchTable.remove(path);
-            if (watchers == null || watchers.isEmpty()) {
-                if (LOG.isTraceEnabled()) {
-                    ZooTrace.logTraceMessage(LOG,
-                            ZooTrace.EVENT_DELIVERY_TRACE_MASK,
-                            "No watchers for " + path);
+            for (String localPath : pathIterator.asIterable()) {
```

Comment:
I'm not sure if we should consider a potential performance impact for tall data trees since we now always iterate all the way up every time a watcher is triggered.